### PR TITLE
Res 977

### PR DESCRIPTION
**Pull request by onejgordon**
Renamed test file for KW2D and added a `new k_winners_test.py` testing the updated scatter-based implementation.

https://ntaresearch.atlassian.net/secure/RapidBoard.jspa?rapidView=60&modal=detail&selectedIssue=RES-977&search=1D

### CODE REVIEW

File: tests/k_winners_test.py [hunk 1]
```diff
@@ -38,210 +38,109 @@ class KWinnersTest(unittest.TestCase):
...
+            n,
             percent_on=0.333,
-            channels=3,
             k_inference_factor=0.5,
             boost_strength=1.0,
             boost_strength_factor=0.5,
             duty_cycle_period=1000,
         )
 
+        # Expect 2 winners per batch (33% of 6)
```
Comment (subutai) [inline tests/k_winners_test.py]: I was initially a bit confused. Shouldn't the number of winners be 0.5 * 0.333 * 6 since k_inference_factor=0.5

Is kw.training True here by default? Maybe we should be more explicit.

Comment (onejgordon) [reply tests/k_winners_test.py]: True, that is not clear. Will set the module explicitly to train()

File: nupic/torch/functions/k_winners.py [hunk 2]
```diff
@@ -99,11 +99,9 @@ def forward(ctx, x, duty_cycles, k, boost_strength):
...
         # Compute an output that contains the values of x corresponding to the top k
         # boosted values
-        res = torch.zeros_like(x)
         topk, indices = boosted.topk(k, sorted=False)
-        for i in range(x.shape[0]):
-            res[i, indices[i]] = x[i, indices[i]]
-
+        mask = torch.zeros_like(x).scatter(-1, indices, 1)
+        res = mask * x
         ctx.save_for_backward(indices)
```
Comment (lscheinkman) [inline nupic/torch/functions/k_winners.py]: maybe we could save the `mask` instead of `indices` so we dont have to rebuild the mask in the backward call

Comment (onejgordon) [reply nupic/torch/functions/k_winners.py]: That makes sense. Can we assume that nothing else currently using this module accesses the saved indices?

Comment (lscheinkman) [reply nupic/torch/functions/k_winners.py]: Yes. Saved indices is only used for backward pass.
